app/shutdownhandler.go

```diff
@@ -0,0 +1,188 @@
+package app
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/arquivei/foundationkit/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// ErrorPolicy specifies what should be done when a handler fails
+type ErrorPolicy int
+
+// ShutdownPriority is used to guide the execution of the shutdown handlers
+// during a graceful shutdown. The shutdown is performed from the higher to the lowest
+// priority
+type ShutdownPriority int
+
+const (
+	// ErrorPolicyWarn prints the error as a warning and continues to the next handler. This is the default.
+	ErrorPolicyWarn ErrorPolicy = iota
+	// ErrorPolicyAbort stops the shutdown process and returns an error
+	ErrorPolicyAbort
+	// ErrorPolicyFatal logs the error as Fatal, it means the application will close immediately
+	ErrorPolicyFatal
+	// ErrorPolicyPanic panics if there is an error
+	ErrorPolicyPanic
+)
+
+// ShutdownFunc is a shutdown function that will be executed when the app is shutting down.
+type ShutdownFunc func(context.Context) error
+
+type shutdownHandler struct {
+	Name     string
+	Timeout  time.Duration
+	Policy   ErrorPolicy
+	Priority ShutdownPriority
+
+	Shutdown ShutdownFunc
+	executed bool
+	err      error
+	mu       sync.Mutex
+	index    int
+}
+
+// Used to set priority automatically by insertion order
+// This is to maintain consistency with the previous queue behavior
+var nextErrorPriority = 0
+var errorPriorityMode = errorPriorityModeUnset
+
+const (
+	errorPriorityModeUnset = iota
+	errorPriorityModeInsertion
+	errorPriorityModeOption
+)
+
+// NewShutdownHandler returns a new
+func NewShutdownHandler(name string, fn ShutdownFunc, options ...interface{}) *shutdownHandler {
+	if name == "" {
+		panic("shutdown handler must have a name")
+	}
+	sh := &shutdownHandler{
+		Name:     name,
+		Shutdown: fn,
+	}
+	for _, o := range options {
+		switch opt := o.(type) {
+		case time.Duration:
+			sh.Timeout = opt
+		case ErrorPolicy:
+			sh.Policy = opt
+		case ShutdownPriority:
+			if errorPriorityMode == errorPriorityModeUnset {
+				errorPriorityMode = errorPriorityModeOption
+			}
+			if errorPriorityMode != errorPriorityModeOption {
+				panic("shutdown handler with priority mixed with shutdown hanlder without priority")
+			}
+
+			sh.Priority = opt
+		default:
+			panic(errors.Errorf("invalid shutdown handler option: [%T]%v", o, o))
+		}
+	}
+
+	// Ensure that we don't mixud settings with and without priority
+	// If no priority is given, the insertion order will be used as priority
+	if sh.Priority == 0 {
```

Comment:
So that means that the Priority `0` is an invalid value? this code panics if someone sets this value as Priority. This behavior should be documented if that's the case.